src/goto-symex/field_sensitivity.cpp

```diff
@@ -0,0 +1,253 @@
+/*******************************************************************\
+
+Module: Field-sensitive SSA
+
+Author: Michael Tautschnig
+
+\*******************************************************************/
+
+#include "field_sensitivity.h"
+
+#include <util/arith_tools.h>
+#include <util/c_types.h>
+#include <util/simplify_expr.h>
+#include <util/std_expr.h>
+
+#include "goto_symex_state.h"
+#include "symex_target.h"
+
+field_sensitivityt::field_sensitivityt(
+  const namespacet &ns,
+  symex_targett &target)
+  : ns(ns), target(target)
+{
+}
+
+void field_sensitivityt::apply(exprt &expr, bool write) const
+{
+  if(!run_apply)
+    return;
+
+  if(expr.id() != ID_address_of)
+    Forall_operands(it, expr)
+      apply(*it, write);
+
+  if(expr.id() == ID_symbol && expr.get_bool(ID_C_SSA_symbol) && !write)
+  {
+    expr = get_fields(ns, to_ssa_expr(expr));
+  }
+  else if(
+    !write && expr.id() == ID_member &&
+    to_member_expr(expr).struct_op().id() == ID_struct)
+  {
+    simplify(expr, ns);
```

Comment:
We have a member expression over a struct expression, so this can be collapsed to just access that member directly. We might ask the question whether this can/does appear at all (as the simplifier should have taken care of this before).